### ISSUE TITLE
How to fix that serial terminal charset problem?

### ISSUE THREAD

**Issue by kendling** [kingToolbox/WindTerm#254]
When I open a serial terminal with /dev/ttyUSB0, WindTerm display incorrect output when receive from serial port.

And the MiniCom can display correct output.

How can I fixit?

![2021-11-07_15-45](https://user-images.githubusercontent.com/3919496/140636983-a6c97a09-1813-444a-bef4-30d566a3e4e5.png)
WindTerm

![2021-11-07_15-48](https://user-images.githubusercontent.com/3919496/140636994-ea7448ba-3aa8-423c-b3b7-8098f111b3f2.png)
MiniCom

**Comment by kingToolbox** [comment]
You only need to simply execute the `Right click menu - Reset - Reset Character Set` menu item to restore the normal display.

Strictly speaking, this is not a bug, but a side effect of complying with `Term` specifications. You can read a more detailed explanation [here](https://github.com/kingToolbox/WindTerm/issues/162#issuecomment-900134015).

I will provide a better solution in the next `WindTerm_2.3.0` version.

**Comment by kendling** [comment]
1. `Reset Character Set` menu can't fix this problem.
2. In this case, the output not have \xe or \xf char.
![2021-11-07_20-14](https://user-images.githubusercontent.com/3919496/140644495-67d58884-60d1-4b3a-b202-7b94bc19aa8d.png)

**Comment by kingToolbox** [comment]
In the first screenshot, there are two black `?` at the starting position, but in the second screenshot, the starting text is `entering kernel init...`. This shows that there should still be hidden characters. Can you collect the debug log for me, if possible, the steps are as follows: 

- Open the `ttyUSB0` session.
- Use the second button on the toolbar to disconnect the session. 
- Check the menu item `Menubar - Mode (7) - Debug Mode`.
- Use the third button on the toolbar to reconnect the session. 
- When the output is complete, uncheck the menu item `Menubar - Mode (7) - Debug Mode`.
- Click the menu item `Right click menu - Log - Open Log Folder`.
- Open the folder `debug` and paste your log file in your comment here. You can zip it before upload.

The log content is in plain text and will not contain any private information. You can use any editor to view it. Thank you.

**Comment by kendling** [comment]
Thanks for your help!

I paste 21st line in here. You can see the full log in attachment.

> [thread 23607][elapsed 3286ms] 
> 
> [Time 2021-11-08 09:45:15.367] [Level 0] Enter Session::bridgeRead
> [thread 23607][elapsed 0ms] 
> [Time 2021-11-08 09:45:15.367] bridge read
> length: 32, content:
> entering kernel init...
> start t
> [thread 23607][elapsed 0ms] 
> [Time 2021-11-08 09:45:15.367] [Level 1] Enter Term::readText
> [thread 23607][elapsed 0ms] 
> [Time 2021-11-08 09:45:15.367] === Receive Text ===
> entering kernel init...
> start t
> [thread 23607][elapsed 0ms] 
> [Time 2021-11-08 09:45:15.367] === Main Screen Text ===
>   row: 0 entering kernel init...********************************************************************************* changed
>         --------------------------------------------------------------------------------------------------------
>   row: 0 columns: 104 valid columns: 23
>   row: 1 start t************************************************************************************************* changed
>         --------------------------------------------------------------------------------------------------------
>   row: 1 columns: 104 valid columns: 7

[WindTerm_log_2021-11-08_09.45.12.zip](https://github.com/kingToolbox/WindTerm/files/7493711/WindTerm_log_2021-11-08_09.45.12.zip)

**Comment by kingToolbox** [comment]
Thank you very much for providing a detailed log. I have analyzed the log and confirmed that the data received on line 64 contains `SO`, which caused `WindTerm` to change the character set. I don't understand why the server needs to send this, but I will look into how to deal with such a situation. I will update the progress here. 

In addition, `Right click menu - Reset - Reset Character Set` should be useful. Although the data before the execution cannot be restored to normal, the data received after the execution should be displayed normally, you can try again. 

Although not recommended, you can also try the following methods to temporarily bypass this problem:

- Open the directory `WindTerm_install_path/terminal/terms`.
- Copy `xterm.term` as `xterm-without-so.term`.
- Use a text editor to open the `xterm-without-so.term` file and delete the line:
> 		| (?<ShiftLockingOne>                            \\x0e) # SO
- Restart WindTerm.
- Open the `Session Property Dialog` of the `ttyUSB0` session and change the `Term` to `xterm-without-so.term`.
- Open the `ttyUSB0` session and everything should be OK.

**Comment by kendling** [comment]
Thanks for your help!

First, `Reset Character Set` not work for me. After click `Reset Character Set`, still not work after reset the hardware for receice new output.

Second, `xterm-without-so.term` is work for me.

**Comment by kendling** [comment]
The screenshot using `without-so.term`
![2021-11-09_22-15](https://user-images.githubusercontent.com/3919496/140940224-a57c37e8-eb3c-413e-9811-a7f79b9cf252.png)

**Comment by kingToolbox** [comment]
Yes, if you reset the hardware, you will receive `SO` again, which will still mess up the characters. Another problem is that `LF` should not be displayed. Did you check `MenuBar - View - Symbols - Line Endings`? I guess not, this is also a problem.

Since `WindTerm_2.2.0` is about to be released, I will solve these problems for you in the `WindTerm_2.3.0_Prerelease` versions as soon as possible. Please stay tuned, thank you.

**Comment by kendling** [comment]
Yes, I selected `MenuBar - View - Symbols - Line Endings` option.

Thanks again.

**Comment by kingToolbox** [comment]
Great, that means it's not a bug. :smile:
